README.md
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/a55402d7c89b4084882c3362427132d8)](https://www.codacy.com/app/juliomrqz/nuxt-netlify?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=juliomrqz/nuxt-netlify&amp;utm_campaign=Badge_Grade)
[![Travis](https://img.shields.io/travis/juliomrqz/nuxt-netlify.svg)](https://travis-ci.org/juliomrqz/nuxt-netlify)
[![David](https://img.shields.io/david/peer/juliomrqz/nuxt-netlify.svg)](https://david-dm.org/juliomrqz/nuxt-netlify?type=peer)
[![David](https://img.shields.io/david/dev/juliomrqz/nuxt-netlify.svg)](https://david-dm.org/juliomrqz/nuxt-netlify?type=dev)
[![version](https://img.shields.io/npm/v/@aceforth/nuxt-netlify.svg)](https://www.npmjs.com/package/@aceforth/nuxt-netlify)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/juliomrqz/nuxt-netlify/develop/LICENSE)

# Nuxt Netlify

Dynamically generate `_headers` and `_redirects` files for Netlify in your Nuxt.js projects.

This module supports the creation of [**redirects**][netlify-redirects] and [**header**][netlify-headers-and-basic-auth] rules for your Netlify site: you can easily configure custom headers, basic auth, redirect instructions and rewrite rules from your _nuxt config file_.

*Read this in other languages: [English][docs], [Español][docs-es]*


## Installation

:warning: `node >= 10` and `nuxt >= 2` are required.


```bash 
npm install --save-dev @aceforth/nuxt-netlify
```

or

```bash 
yarn add --dev @aceforth/nuxt-netlify
```

Add `@aceforth/nuxt-netlify` to the `buildModules` section of `nuxt.config.js`:

:warning: If you are using Nuxt `< 2.9.0`, use `modules` instead. 

```js
{
  buildModules: [
    '@aceforth/nuxt-netlify',
  ],

  netlify: { 
    mergeSecurityHeaders: true 
  }
}
```

or 


```js
{
  buildModules: [
    [
      '@aceforth/nuxt-netlify',
      {
        mergeSecurityHeaders: true
      }
    ]
  ]
}
```


## Usage

The default configuration will generate an empty `_redirects` file and a `_headers` file with some caching and security headers:

```text
# _headers

/*
  Referrer-Policy: origin
  X-Content-Type-Options: nosniff
  X-Frame-Options: DENY
  X-XSS-Protection: 1; mode=block

/_nuxt/*
  Cache-Control: public, max-age=31536000, immutable

/sw.js
  Cache-Control: no-cache
```

> :warning: the `/_nuxt/*` reference automatically changes with the value of [`build.publicPath`][nuxt-docs-build-publicPath].

### Headers

The headers object represents a JS version of the [Netlify `_headers` file format][netlify-headers-and-basic-auth]. You should pass in a object with string keys (representing the paths) and an array of strings for each header. For example:


You can add extra headers as follows:

```js
const pkg = require('./package.json')

{
  netlify: { 
    headers: {
      '/*': [
        'Access-Control-Allow-Origin: *',
        `X-Build: ${pkg.version}`
      ],
      '/favicon.ico': [
        'Cache-Control: public, max-age=86400'
      ]
    }
  }
}
```

that will generate:

```text
# _headers

/*
  Referrer-Policy: origin
  X-Content-Type-Options: nosniff
  X-Frame-Options: DENY
  X-XSS-Protection: 1; mode=block
  Access-Control-Allow-Origin: *
  X-Build: 1.0.1

/_nuxt/*
  Cache-Control: public, max-age=31536000, immutable

/sw.js
  Cache-Control: no-cache
  
/favicon.ico
  Cache-Control: public, max-age=86400
```

### Redirects

You can also add [redirects][netlify-redirects], as many as you like. You should pass in an array of objects with the redirection attributes. For example:


```js
{
  netlify: { 
    redirects: [
      {
        from: '/home',
        to: '/'
      },
      {
        from: '/my-redirect',
        to: '/',
        status: 302,
        force: true
      },
      {
        from: '/en/*',
        to: '/en/404.html',
        status: 404
      },
      {
        from: '/*',
        to: '/index.html',
        status: 200
      },
      {
        from: '/store',
        to: '/blog/:id',
        query: {
          id: ':id'
        }
      },
      {
        from: '/',
        to: '/china',
        status: 302,
        conditions: {
          Country: 'cn,hk,tw'
        }
      }
    ]
  }
}
```

will generate:

```text
# _redirects

/home               /               301
/my-redirect        /               302!
/en/*               /en/404.html    404
/*                  /index.html     200
/store    id=:id    /blog/:id       301
/                   /china          302    Country=cn,hk,tw
```


See the [configuration][docs-configuration] section for all available options.

## Documentation & Support

- 📄 If you want extra details of how to configure and use this project, the **full documentation** is available at [https://marquez.co/docs/nuxt-netlify/][docs].
- 🐞 For **Bug reports** or **Feature requests**, use the [Issues section][issues].
- 💬 For **questions**, you can also use the [Issues section][issues].
- 🚀 You may also want to **follow me** [on Twitter][twitter].

## Professional Support

This project is sponsored by me, a Full Stack Developer. If you require Professional Assistance on your project(s), please contact me at [https://marquez.co][contact-page].

## Code of Conduct

Everyone participating in this project is expected to agree to abide by the [Code of Conduct][code-of-conduct].


## License

Code released under the [MIT License][license-page].


![](https://ga-beacon.appspot.com/UA-65885578-17/juliomrqz/nuxt-netlify?pixel)

[docs]: https://marquez.co/docs/nuxt-netlify/?utm_source=github&utm_medium=readme&utm_campaign=nuxt-netlify
[docs-es]: https://marquez.co/es/docs/nuxt-netlify/?utm_source=github&utm_medium=readme&utm_campaign=nuxt-netlify
[docs-configuration]: https://marquez.co/docs/nuxt-netlify/configuration/?utm_source=github&utm_medium=readme&utm_campaign=nuxt-netlify
[nuxt-docs-build-publicPath]: https://nuxtjs.org/api/configuration-build#publicPath
[netlify-headers-and-basic-auth]: https://www.netlify.com/docs/headers-and-basic-auth/
[netlify-redirects]: https://www.netlify.com/docs/redirects/
[issues]: https://github.com/juliomrqz/nuxt-netlify/issues
[twitter]: https://twitter.com/juliomrqz
[contact-page]: https://marquez.co?utm_source=github&utm_medium=readme&utm_campaign=nuxt-netlify
[code-of-conduct]: https://www.contributor-covenant.org/version/2/0/code_of_conduct/
[license-page]: https://github.com/juliomrqz/nuxt-netlify/blob/develop/LICENSE


CHANGELOG.md
# [1.1.0](https://github.com/juliomrqz/nuxt-netlify/compare/v1.0.0...v1.1.0) (2020-07-19)


* dependencies updates

### Refactor

* update layout of docs articles ([08ea91a](https://github.com/juliomrqz/nuxt-optimized-images/commit/08ea91a4472654d764d83f01d52d7a16ff2ac82b))
* update project owner ([18216ad](https://github.com/juliomrqz/nuxt-optimized-images/commit/18216ad0cb2592770db6f4cb08f0d41815e8d154))


# [1.0.0](https://github.com/juliomrqz/nuxt-netlify/compare/v0.2.0...v1.0.0) (2020-04-12)


* refactor!: update parent company ([3a93e72](https://github.com/juliomrqz/nuxt-netlify/commit/3a93e72a4df48b266e1fe23f3762485fa7d1331b))


### BREAKING CHANGES

* **The package has been renamed from `@bazzite/nuxt-netlify` to `@aceforth/nuxt-netlify`.**

To upgrade

1. `npm install --save-dev @aceforth/nuxt-netlify` 

   or `yarn add --dev @aceforth/nuxt-netlify`

2. `npm uninstall @bazzite/nuxt-netlify` 

   or `yarn remove @bazzite/nuxt-netlify`

3. replace:

```js
{
  buildModules: [
    '@bazzite/nuxt-netlify',
  ],
}
```

with

```js
{
  buildModules: [
    '@aceforth/nuxt-netlify',
  ],
}
```


That’s it, there are no functional changes compared to `@bazzite/nuxt-netlify@0.2.0`.



# [0.2.0](https://github.com/juliomrqz/nuxt-netlify/compare/v0.1.1...v0.2.0) (2020-03-31)


### Bug Fixes

* **docs:** update Installation instructions ([d0fbe6f](https://github.com/juliomrqz/nuxt-netlify/commit/d0fbe6fb1b41ac861ca43f4930b62b7c1c6d9f24))


* refactor!: drop support for Node.js 8 ([34075b7](https://github.com/juliomrqz/nuxt-netlify/commit/34075b7b54dd0a71b5756ce7ecc9a9a9e0c5cd1d))


### Features

* add support for node v12 on travis tests ([21b43c0](https://github.com/juliomrqz/nuxt-netlify/commit/21b43c03e3ceb78a6667374841bb23a6c6620c2d))


### BREAKING CHANGES

* minimum required Node.js version is 10.x



## [0.1.1](https://github.com/juliomrqz/nuxt-netlify/compare/v0.1.0...v0.1.1) (2019-06-12)

* **docs:** extend the description of the project ([21f5e09](https://github.com/juliomrqz/nuxt-netlify/commit/21f5e09))


# 0.1.0 (2019-03-14)

First stable release


test/headers.test.js
const { appendHeaders, createHeadersContent } = require('../lib/utils')

describe('headers', () => {
  it('can append headers', () => {
    const headers1 = {
      '/*': [
        'X-Frame-Options: DENY',
        'X-XSS-Protection: 1; mode=block',
        'X-Content-Type-Options: nosniff'
      ]
    }

    const headers2 = {
      '/*': [
        'X-UA-Compatible: ie=edge',
        'Access-Control-Allow-Origin: *',
        'Referrer-Policy: origin'
      ]
    }

    const headers3 = {
      '/*': [
        'X-Frame-Options: SAMEORIGIN'
      ]
    }

    expect(appendHeaders(headers1, '/*', headers2['/*'])).toMatchSnapshot()
    expect(appendHeaders(headers1, '/*', headers3['/*'])).toMatchSnapshot()
  })

  it('can create header content', () => {
    const headers1 = [
      'Cache-Control: public, max-age=31536000, immutable'
    ]

    const headers2 = [
      'X-Frame-Options: DENY',
      'X-XSS-Protection: 1; mode=block'
    ]

    const headers3 = [
      'Basic-Auth: someuser:somepassword anotheruser:anotherpassword'
    ]

    const headers4 = [
      'Link: </style.css>; rel=preload; as=style',
      'Link: </main.js>; rel=preload; as=script',
      'Link: </image.jpg>; rel=preload; as=image'
    ]

    expect(appendHeaders(createHeadersContent(headers1, '/static/*'))).toMatchSnapshot()
    expect(appendHeaders(createHeadersContent(headers2, '/templates/index.html'))).toMatchSnapshot()
    expect(appendHeaders(createHeadersContent(headers3, '/something/*'))).toMatchSnapshot()
    expect(appendHeaders(createHeadersContent(headers4, '/*'))).toMatchSnapshot()
  })
})


test/redirects.test.js
const { createRedirectContent } = require('../lib/utils')

describe('redirects', () => {
  it('can generate a simple redirect', () => {
    expect(createRedirectContent({
      from: '/home',
      to: '/'
    })).toEqual('/home    /    301\n')

    expect(createRedirectContent({
      from: '/blog/my-post.php',
      to: '/blog/my-post'
    })).toEqual('/blog/my-post.php    /blog/my-post    301\n')

    expect(createRedirectContent({
      from: '/google',
      to: 'https://www.google.com'
    })).toEqual('/google    https://www.google.com    301\n')

    expect(createRedirectContent({
      from: '/news/*',
      to: '/blog/:splat'
    })).toEqual('/news/*    /blog/:splat    301\n')
  })

  it('can generate a redirect with a different status code', () => {
    expect(createRedirectContent({
      from: '/my-redirect',
      to: '/',
      status: 302
    })).toEqual('/my-redirect    /    302\n')

    expect(createRedirectContent({
      from: '/pass-through',
      to: '/index.html',
      status: 200
    })).toEqual('/pass-through    /index.html    200\n')

    expect(createRedirectContent({
      from: '/ecommerce',
      to: '/store-closed',
      status: 404
    })).toEqual('/ecommerce    /store-closed    404\n')

    expect(createRedirectContent({
      from: '/en/*',
      to: '/en/404.html',
      status: 404
    })).toEqual('/en/*    /en/404.html    404\n')

    expect(createRedirectContent({
      from: '/*',
      to: '/index.html',
      status: 200
    })).toEqual('/*    /index.html    200\n')
  })

  it('can force redirection', () => {
    expect(createRedirectContent({
      from: 'http://blog.yoursite.com/*',
      to: 'https://www.yoursite.com/blog/:splat',
      force: true
    })).toEqual('http://blog.yoursite.com/*    https://www.yoursite.com/blog/:splat    301!\n')

    expect(createRedirectContent({
      from: 'http://blog.yoursite.com/*',
      to: 'https://www.yoursite.com/blog/:splat',
      status: 302,
      force: true
    })).toEqual('http://blog.yoursite.com/*    https://www.yoursite.com/blog/:splat    302!\n')
  })

  it('can generate a redirect with query params', () => {
    expect(createRedirectContent({
      from: '/store',
      to: '/blog/:id',
      query: {
        id: ':id'
      }
    })).toEqual('/store    id=:id    /blog/:id    301\n')

    expect(createRedirectContent({
      from: '/articles',
      to: '/posts/:tag/:id',
      query: {
        id: ':id',
        tag: ':tag'
      }
    })).toEqual('/articles    id=:id  tag=:tag    /posts/:tag/:id    301\n')
  })

  it('can generate a redirect with conditons', () => {
    expect(createRedirectContent({
      from: '/',
      to: '/china',
      status: 302,
      conditions: {
        Country: 'cn,hk,tw'
      }
    })).toEqual('/    /china    302    Country=cn,hk,tw\n')

    expect(createRedirectContent({
      from: '/',
      to: '/israel',
      status: 302,
      conditions: {
        Country: 'il'
      }
    })).toEqual('/    /israel    302    Country=il\n')

    expect(createRedirectContent({
      from: '/admin/*',
      status: 200,
      force: true,
      conditions: {
        Role: 'admin'
      }
    })).toEqual('/admin/*    200!    Role=admin\n')

    expect(createRedirectContent({
      from: '/private/*',
      status: 200,
      force: true,
      conditions: {
        Role: 'editor,admin'
      }
    })).toEqual('/private/*    200!    Role=editor,admin\n')
  })
})


test/file-headers.test.js
const { Map } = require('immutable')

const { DEFAULT_CONFIG } = require('../lib/constants')
const generateHeaders = require('../lib/headers')

const defaultOptions = Map(Object.assign({}, DEFAULT_CONFIG, { publicPath: '/_nuxt/' }))
const extraHeaders = {
  '/templates/index.html': [
    'X-Frame-Options: DENY',
    'X-XSS-Protection: 1; mode=block'
  ],
  '/templates/index2.html': [
    'X-Frame-Options: SAMEORIGIN'
  ],
  '/something/*': [
    'Basic-Auth: someuser:somepassword anotheruser:anotherpassword'
  ]
}

describe('file _headers', () => {
  it('can generate default file', () => {
    const options = defaultOptions.toJS()
    const content = generateHeaders(options)
    expect(content).toMatchSnapshot()
  })

  it('can generate default file with a public path as url', () => {
    const options = defaultOptions.set('publicPath', 'https://cdn.example.com').toJS()
    const content = generateHeaders(options)
    expect(content).toMatchSnapshot()
  })

  it('can disable security headers', () => {
    const options = defaultOptions.set('mergeSecurityHeaders', false).toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })

  it('can disable caching headers', () => {
    const options = defaultOptions.set('mergeCachingHeaders', false).toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })

  it('can disable security & caching headers', () => {
    const options = defaultOptions
      .set('mergeSecurityHeaders', false)
      .set('mergeCachingHeaders', false)
      .toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })

  it('can sort headers alphabetically', () => {
    const options = defaultOptions
      .set('transformHeaders', (headers, path) => headers.reverse())
      .toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })

  it('can add extra headers', () => {
    const options = defaultOptions
      .set('headers', extraHeaders)
      .toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })

  it('can add extra headers with no defaults', () => {
    const options = defaultOptions
      .set('mergeSecurityHeaders', false)
      .set('mergeCachingHeaders', false)
      .set('headers', extraHeaders)
      .toJS()
    const content = generateHeaders(options)

    expect(content).toMatchSnapshot()
  })
})


lib/index.js
/**
 * Copyright (c) 2020, Julio Marquez
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const path = require('path')
const consola = require('consola')

const { DEFAULT_CONFIG } = require('./constants')
const generateHeaders = require('./headers')
const generateRedirects = require('./redirects')
const { createFile } = require('./utils')

const logger = consola.withScope('nuxt:@aceforth/nuxt-netlify')

function nuxtOptimizedImages (moduleOptions) {
  const { rootDir, generate, build } = this.options
  const options = Object.assign({}, DEFAULT_CONFIG, this.options.netlify, moduleOptions)
  options.publicPath = build.publicPath

  // validate options
  if (typeof options.headers !== 'object') {
    logger.fatal('The `headers` property must be an object')
  } else if (!Array.isArray(options.redirects)) {
    logger.fatal('The `redirects` property must be an array')
  } else if (typeof options.transformHeaders !== 'function') {
    logger.fatal('The `transformHeaders` property must be a function')
  }

  // Generate headers and redirects in dist
  this.nuxt.hook('generate:done', async () => {
    // generate headers
    try {
      const headersPath = path.resolve(rootDir, generate.dir, '_headers')
      const headersContent = generateHeaders(options)
      await createFile(headersPath, headersContent)
      logger.success('Generated /_headers')
    } catch (error) {
      logger.error(error)
    }

    // generate redirects
    try {
      const redirectsPath = path.resolve(rootDir, generate.dir, '_redirects')
      const redirectsContent = generateRedirects(options)
      await createFile(redirectsPath, redirectsContent)
      logger.success('Generated /_redirects')
    } catch (error) {
      logger.error(error)
    }
  })
}

module.exports = nuxtOptimizedImages
module.exports.meta = require('../package.json')


lib/utils.js
const fs = require('fs-extra')
const { DEFAULT_HEADER_VALUES } = require('./constants')

const isEmptyObject = input => Object.keys(input).length === 0

const createFile = async (filePath, content) => {
  if (await fs.exists(filePath)) {
    await fs.appendFile(filePath, `\n\n${content}`)
  } else {
    await fs.ensureFile(filePath)
    await fs.writeFile(filePath, content)
  }
}

const createHeadersContent = (headers, path) => {
  let content = `${path}\n`

  headers.forEach(h => {
    content += `  ${h}\n`
  })

  return `${content}\n`
}

const appendHeaders = (headersList, key, headers) => {
  if (headersList[key]) {
    headersList[key].push(...headers)
  } else {
    headersList[key] = headers
  }

  headersList[key] = Array.from(new Set(headersList[key]))

  return headersList
}

const isUrl = (url) => {
  return ['http', '//'].some(str => url.startsWith(str))
}

const createRedirectContent = r => {
  const redirect = Object.assign({}, DEFAULT_HEADER_VALUES, r)
  const divider = '    '

  // from
  let content = `${redirect.from}${divider}`

  // query params
  if (redirect.query && !isEmptyObject(redirect.query)) {
    content += Object.keys(redirect.query)
      .map(k => `${k}=${redirect.query[k]}`)
      .join('  ')
    content += divider
  }

  // to
  if (redirect.to) {
    content += `${redirect.to}${divider}`
  }

  // status
  content += `${redirect.status}${redirect.force ? '!' : ''}`

  // conditions
  if (redirect.conditions && !isEmptyObject(redirect.conditions)) {
    content += divider
    content += Object.keys(redirect.conditions)
      .map(k => `${k}=${redirect.conditions[k]}`)
      .join('  ')
  }

  return `${content}\n`
}

module.exports = {
  createFile,
  createHeadersContent,
  appendHeaders,
  isUrl,
  createRedirectContent
}


lib/headers.js
const { SECURITY_HEADERS, CACHING_HEADERS, FILE_COMMENT } = require('./constants')
const { createHeadersContent, appendHeaders, isUrl } = require('./utils')

module.exports = (options) => {
  const { transformHeaders, publicPath } = options
  let content = FILE_COMMENT

  if (options.mergeSecurityHeaders) {
    options.headers = appendHeaders(options.headers, '/*', SECURITY_HEADERS['/*'])
  }

  if (options.mergeCachingHeaders) {
    // public path
    if (!isUrl(publicPath)) {
      options.headers = appendHeaders(options.headers,
        `${publicPath}*`,
        CACHING_HEADERS['/_nuxt/*'])
    }

    // sw.js
    options.headers = appendHeaders(options.headers, '/sw.js', CACHING_HEADERS['/sw.js'])
  }

  // transform and generate headers content
  content += Object.keys(options.headers)
    .map(k => createHeadersContent(transformHeaders(options.headers[k], k), k))
    .join('')

  return content
}


lib/constants.js
/**
 * Plugin values
 */
const DEFAULT_CONFIG = {
  // add more headers
  headers: {},
  // turn off the default security headers
  mergeSecurityHeaders: true,
  // turn off the default caching headers
  mergeCachingHeaders: true,
  // optional transform for manipulating headers under each path (e.g.sorting), etc.
  transformHeaders: (headers, path) => headers,
  // add redirects
  redirects: []
}

const DEFAULT_HEADER_VALUES = {
  from: null,
  to: null,
  status: 301,
  force: false,
  query: {},
  conditions: {}
}

const SECURITY_HEADERS = {
  '/*': [
    'Referrer-Policy: origin',
    'X-Content-Type-Options: nosniff',
    'X-Frame-Options: DENY',
    'X-XSS-Protection: 1; mode=block'
  ]
}

const CACHING_HEADERS = {
  '/_nuxt/*': ['Cache-Control: public, max-age=31536000, immutable'],
  '/sw.js': ['Cache-Control: no-cache']
}

const FILE_COMMENT = '## Created with https://github.com/juliomrqz/nuxt-netlify\n'

module.exports = {
  DEFAULT_CONFIG,
  DEFAULT_HEADER_VALUES,
  SECURITY_HEADERS,
  CACHING_HEADERS,
  FILE_COMMENT
}


example/nuxt.config.js
const path = require('path')
const pkg = require('./package')

module.exports = {
  mode: 'universal',

  /*
  ** Headers of the page
  */
  head: {
    title: pkg.name,
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'description', name: 'description', content: pkg.description }
    ],
    link: [
      { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' }
    ]
  },

  /*
  ** Customize the progress-bar color
  */
  loading: { color: '#fff' },

  /*
  ** Global CSS
  */
  css: [
  ],

  /*
  ** Plugins to load before mounting the App
  */
  plugins: [
  ],

  /*
  ** Nuxt.js modules
  */
  modules: [
    path.resolve('../lib/index.js')
  ],

  /*
  ** Build configuration
  */
  build: {
    /*
    ** You can extend webpack config here
    */
    extend(config, ctx) {

    }
  },
  netlify: {
    redirects: [
      {
        from: 'https://example.com/temp/*',
        to: 'https://example.com/:splat*'
      },
      {
        from: '/en/*',
        to: '/en/404.html',
        status: 404
      },
      {
        from: '/articles',
        to: '/posts/:tag/:id',
        status: 301,
        force: true,
        query: {
          id: ':id',
          tag: ':tag',
        }
      },
      {
        from: '/china/*',
        to: '/china/zh-cn/:splat',
        status: 302,
        conditions: {
          Language: 'zh',
          Country: 'cn,hk,tw',
        }
      }
    ]
  }
}
